internal/controller/pkg/manager/reconciler.go

```diff
@@ -407,6 +522,13 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		r.record.Event(p, event.Warning(reasonInstall, errors.New(errUnknownPackageRevisionHealth)))
 	}
 
+	if pr.GetUID() == "" && imageConfig != "" {
+		// We only record this event if the revision is new, as we don't want to
+		// spam the user with events if the revision already exists.
```

Comment:
Today we need pull secrets in 3 different places: 

- Package reconciler - to fetch the head/description of the image just to calculate the hash and create a revision (if it is not there)
- PackageRevision reconciler - to pull actual image layers into the cache and apply resources. It also passes the pull secret to provider deployment so that it can pull its own image.
- Lock reconciler - to list the available tags to find a version that satisfies dependencies.

We are querying a pull secret matching from an image in all three places above, it is not like, we pickup a pull secret in package reconciler and somehow use it in other reconcilers forever.

> I could also imagine an edge case of a package that we previously were able to pull without credentials being updated to require credentials in future. We'd fail to pull and need an ImageConfig if the pod restarted and lost its cache (or was just set to pull always).

I believe the current implementation would be immune to such a scenario. Assuming we created a new image config after the repo is updated to be private, the package reconciler will query and pick up a pull secret for this image and will fetch the image description. If the hash of the image stays same, we will not create a new revision, hence not throw a new event. But, the package revision reconciler will get the pull secret and download no longer existing image into the cache. 

> What about (in future) validation? I could imagine that might want to apply retroactively.

I believe we need to handle case by base. For image validations, I am planning to have a dedicated condition managed by a separate reconciler (as I mentioned in the design).